### PR TITLE
feat: add streamChunkMut

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

will be used in https://github.com/risingwavelabs/risingwave/pull/11070. all the rowMut will be in a hashMap and do some modifications on EachRow. Finally, the hashMap will be dropped and StreamChunk will be built from the StreamChunkMut.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/array/stream_chunk.rs [hunk 1]
```diff
@@ -306,6 +307,102 @@ impl EstimateSize for StreamChunk {
...
+impl StreamChunkMut {
+    pub fn set_vis(&mut self, n: usize, val: bool) {
+        self.vis.set(n, val);
+    }
+
+    pub fn set_op(&mut self, n: usize, val: Op) {
+        self.ops.set(n, val);
+    }
+
+    pub unsafe fn to_mut_rows(&self) -> impl Iterator<Item = OpRowMutRef<'_>> {
```
Comment (wangrunji0408) [inline src/common/src/array/stream_chunk.rs:398]: I feel this function can be safe?
```suggestion
    pub fn iter_rows_mut(&mut self) -> impl Iterator<Item = OpRowMutRef<'_>> {
```

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

**Comment by BugenZhao** [comment]
Would you please elaborate more about the motivation in the PR's body?

File: src/common/src/array/stream_chunk.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/array/stream_chunk.rs:398]: BTW, can we get rid of `unsafe` by manually implementing the iterator?

File: src/common/src/array/stream_chunk.rs [hunk 2]
```diff
@@ -306,6 +307,102 @@ impl EstimateSize for StreamChunk {
...
+}
+
+impl OpsMut {
+    pub fn new(ops: Arc<[Op]>) -> Self {
+        Self {
+            state: OpsMutState::ArcRef(ops),
+        }
+    }
+
+    pub fn set(&mut self, n: usize, val: Op) {
```
[the remark this answers is not in the document]
Comment (st1page) [reply src/common/src/array/stream_chunk.rs:327]: Not exactly, the copy will be happening every time `make_mut` is called. I need a physical variable that can be stored somewhere and `&mut T` is not enough.

File: src/common/src/array/stream_chunk.rs (hunk 1 above)
Comment (st1page) [reply src/common/src/array/stream_chunk.rs:398]: :thinking:

File: src/common/src/array/stream_chunk.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (st1page) [reply src/common/src/array/stream_chunk.rs:327]: Oh, learned so much, the clone won't happen when no other Arc pointers shares. But I have finished the current implementation which saves some atomic operations.... :innocent:

File: src/common/src/array/stream_chunk.rs [hunk 3]
```diff
@@ -306,6 +307,102 @@ impl EstimateSize for StreamChunk {
...
+        if let OpsMutState::MutRef(v) = &mut self.state {
+            v[n] = val;
+        } else {
+            let state = mem::replace(&mut self.state, OpsMutState::Undefined); // intermediate state
```
[the remark this answers is not in the document]
Comment (st1page) [reply src/common/src/array/stream_chunk.rs:331]: :innocent: learned so much

File: src/common/src/array/stream_chunk.rs [hunk 4]
```diff
@@ -306,6 +307,102 @@ impl EstimateSize for StreamChunk {
     }
 }
 
+enum OpsMutState {
+    ArcRef(Arc<[Op]>),
+    MutRef(Vec<Op>),
```
[the remark this answers is not in the document]
Comment (st1page) [reply src/common/src/array/stream_chunk.rs]: fixed 
https://github.com/risingwavelabs/risingwave/pull/11808/commits/098745366339fbead95b6d1f451bb585627bd09e

File: src/common/src/array/stream_chunk.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/common/src/array/stream_chunk.rs:400]: > I guess it's safe only if we bound the lifetime of returning iterator.

The iterator's items are borrowing `self` so `+ '_` is not needed. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f253acc4590bfd3de4a619be201c65b

Comment (wangrunji0408) [reply src/common/src/array/stream_chunk.rs:400]: > BTW, can we get rid of unsafe by manually implementing the iterator?

Each item has a `&mut` to the chunk and multiple items can exist at the same time. So I guess it's impossible to implement in safe Rust.

Comment (st1page) [reply src/common/src/array/stream_chunk.rs]: fixed.

**Pull request merged**